src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -203,42 +204,45 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
     else super.++:(that)
   }
-  private def addSimple[C >: (A, B), That](that: TraversableOnce[C]): That = {
+  private def addSimple[C >: (A, B), That](that: TraversableOnce[C])(implicit bf: CanBuildFrom[HashMap[A, B], C, That]): That = {
     //here we know that That =:= HashMap[_, _], or compatible with it
-    if (this eq that.asInstanceOf[AnyRef]) that.asInstanceOf[That]
-    else if (that.isEmpty) this.asInstanceOf[That]
-    else that match {
-      case thatHash: HashMap[A, B] =>
-        val merger: Merger[A, B] = HashMap.liftMerger[A, B](null)
-        // merger prefers to keep than replace
-        // so we invert
-        (this.merge0(thatHash, 0, merger.invert)).asInstanceOf[That]
-
-      case that:HasForeachEntry[A, B] =>
-        object adder extends Function2[A, B, Unit] {
-          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
-          val merger = HashMap.liftMerger[A, B](null)
-
-          override def apply(key: A, value: B): Unit = {
-            result = result.updated0(key, computeHash(key), 0, value, null, merger)
+    if (this eq that.asInstanceOf[AnyRef]) castToThat(that)
+    else if (that.isEmpty) castToThat(this)
+    else {
+      val merger = HashMap.concatMerger[A, B].invert
+      val result: HashMap[A, B] = that match {
+        case thatHash: HashMap[A, B] =>
+          this.merge0(thatHash, 0, HashMap.concatMerger[A, B].invert)
+
+        case that:HasForeachEntry[A, B] =>
+          object adder extends Function2[A, B, Unit] {
+            var result: HashMap[A, B] = HashMap.this
+            override def apply(key: A, value: B): Unit = {
+              result = result.updated0(key, computeHash(key), 0, value, null, merger)
+            }
           }
-        }
-        that foreachEntry adder
-        adder.result.asInstanceOf[That]
-      case that =>
-        object adder extends Function1[(A,B), Unit] {
-          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
-          val merger = HashMap.liftMerger[A, B](null)
-
-          override def apply(kv: (A, B)): Unit = {
-            val key = kv._1
-            result = result.updated0(key, computeHash(key), 0, kv._2, kv, merger)
+          that foreachEntry adder
+          adder.result
+        case that =>
+          object adder extends Function1[(A,B), Unit] {
+            var result: HashMap[A, B] = HashMap.this
+            override def apply(kv: (A, B)): Unit = {
+              val key = kv._1
+              result = result.updated0(key, computeHash(key), 0, kv._2, kv, merger)
+            }
           }
-        }
-        that.asInstanceOf[scala.Traversable[(A,B)]] foreach adder
-        adder.result.asInstanceOf[That]
+          that.asInstanceOf[scala.Traversable[(A,B)]] foreach adder
+          adder.result
+      }
+      castToThat(result)
     }
   }
+  private[this] def castToThat[C, That](m: HashMap[A, B])(implicit bf: CanBuildFrom[HashMap[A, B], C, That]): That = {
```

Comment:
this method is just to hide the cast?
Can we make it @inline at least, and drop the cbf

Reply:
I contend its fine to rely on JIT here.